### ISSUE TITLE
improve "X out of Y instance types were excluded because they would breach provisioner limits"

### ISSUE THREAD

**Issue by grosser** [aws/karpenter#2478]
**Tell us about your request**
```
329 out of 550 instance types were excluded because they would breach provisioner limits
```
atm this is misleading and noisy since I have
```
  - key: "node.kubernetes.io/instance-type"
    operator: In
    values: ["m6i.xlarge"]
```
so it would be nice to apply these filters first and then log if any of the possible nodes got excluded, so I have an actionable "I need to check my type list" log message

**Are you currently working around this issue?**
Filter out these misleading logs.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by spring1843** [comment]
The log is being generated [here](https://github.com/aws/karpenter/blob/de309caa29a76aa3e8cd220c525f9b3970282c10/pkg/controllers/provisioning/scheduling/scheduler.go#L213).

> so it would be nice to apply these filters first and then log if any of the possible nodes got excluded, so I have an actionable "I need to check my type list" log message

Are you saying for each instance type there should be a log saying why it was excluded? Can you provide an example of what ideal logs would look like?

**Comment by grosser** [comment]
no, it should first filter out all the types that are not possible (arm, or not in requested instance list) and then try to fit instances by their size.
That way this message would usually be empty and if it is not empty it would be a good warning sign that we reached max capacity.
Whereas atm it's mostly spam since there are always some instances filtered out (since we pick instance types and don't run on arm for example) so it will always say "329 out of 550" or "328 out of 550" and is not very helpful.

**Comment by njtran** [comment]
Hey @grosser, sorry for the very delayed response. We made some changes to add in visibility about when provisioning requirements don't match in https://github.com/aws/karpenter-core/pull/317, if this is any interest to you.

For filtering out instance types we actually do it in this order, which is what you're proposing we switch:
1) filter out instance types based on provisioner limits
2) filter out remaining instance types by requirements

We've often seen that requirements-based incompatibilities are much more verbose due to the cardinality of possibilities a user can define. So, if we filter out the instance types that don't match limits, it'll reduce the complexity of figuring incompatibilities out. If the order is reversed, you'll see logs that show smaller numbers in the log line you're proposing, but then when you remove those instance types from your requirements, you could find that those were the only instance types that met requirements. 

In all, if you're near provisioner limits, you'll see a higher number in this log line, but if you're further away from breaching limits, you may see a longer list of instance types to try and find incompatibilities with. I'm on the side of erring to give more information for requirements (keeping the order as is) as resource limits comparison can be figured out by looking at your [existing resources in the provisioner status](https://github.com/aws/karpenter-core/blob/main/pkg/apis/v1alpha5/provisioner_status.go#L36), and comparing instance type sizes to that. 

Maybe as a middle ground, we could print the first 10 or 20 instance types we don't see as matching resource limits, if that would help you filter out which ones you need to remove.

**Comment by grosser** [comment]
yeah a limited list followed by `(23 more)` could be nice
